pkg/trace/info/stats.go

```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
 	metrics.Count("datadog.trace_agent.receiver.events_extracted", eventsExtracted, tags, 1)
 	metrics.Count("datadog.trace_agent.receiver.events_sampled", eventsSampled, tags, 1)
 	metrics.Count("datadog.trace_agent.receiver.payload_accepted", requestsMade, tags, 1)
+
+	for reason, count := range ts.TracesDropped.toMap() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_dropped", count, append(tags, "reason:"+reason), 1)
+	}
+
+	for reason, count := range ts.TracesMalformed.toMap() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_malformed", count, append(tags, "reason:"+reason), 1)
+	}
+}
+
+// statsStructToMap marshals the provided struct into a map[string]int64 by first marshaling it into json
+// then marshing from json into the map
+func statsStructToMap(v interface{}, name string) (result map[string]int64) {
+	statsB, err := json.Marshal(&v)
+	if err != nil {
+		log.Errorf("Failed to marshal %s to json: %s", name, err)
+		return result
+	}
+	err = json.Unmarshal(statsB, &result)
+	if err != nil {
+		log.Errorf("Failed to unmarshal %s json to map: %s", name, err)
+	}
+	return result
+}
+
+// inlineStatsMap serializes the entries in this map into format "key1: value1, key2: value2, ...", sorted by key to
+// ensure output order is always the same for the same keys
+func inlineStatsMap(statsMap map[string]int64) string {
+	keys := make([]string, len(statsMap))
+	i := 0
+	for k := range statsMap {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+
+	results := make([]string, len(statsMap))
+	for i, key := range keys {
+		results[i] = fmt.Sprintf("%s:%d", key, statsMap[key])
+	}
+
+	return strings.Join(results, ", ")
+}
+
+// TracesDroppedStats contains counts for reasons traces have been dropped
+type TracesDroppedStats struct {
+	DecodingError int64 `json:"decoding_error,omitempty"`
+	EmptyTrace    int64 `json:"empty_trace,omitempty"`
+	TraceIDZero   int64 `json:"trace_id_zero,omitempty"`
+	SpanIDZero    int64 `json:"span_id_zero,omitempty"`
+	ForeignSpan   int64 `json:"foreign_span,omitempty"`
+}
+
+// AtomicCopy returns a safe copy of the struct by doing atomic reads from all fields
+func (s *TracesDroppedStats) AtomicCopy() (result TracesDroppedStats) {
+	result.DecodingError = atomic.LoadInt64(&s.DecodingError)
+	result.EmptyTrace = atomic.LoadInt64(&s.EmptyTrace)
+	result.TraceIDZero = atomic.LoadInt64(&s.TraceIDZero)
+	result.SpanIDZero = atomic.LoadInt64(&s.SpanIDZero)
+	result.ForeignSpan = atomic.LoadInt64(&s.ForeignSpan)
+	return result
+}
+
+func (s *TracesDroppedStats) toMap() (result map[string]int64) {
+	return statsStructToMap(s.AtomicCopy(), "TracesDroppedStats")
+}
+
+func (s *TracesDroppedStats) String() string {
```

Comment:
Here we can use `strings.Builder` to do something (again verbose) like:
```go
func (s *TracesDroppedStats) String() string {
    var str strings.Builder
    for tag, value := range map[string]int64{
        tagReasonDecodingError: atomic.LoadInt64(&s.DecodingError),
        tagReasonEmptyTrace: atomic.LoadInt64(&s.EmptyTrace),
        // ...
    } {
        if value > 0 {
            str.WriteString(tag)
            str.WriteByte(':')
            str.WriteString(strconv.Itoa(valut))
            str.WriteByte(' ')
        }
    }
    return str
}
```
This is again super-verbose, but IMO much clearer. Generally, this is the preferred approach in Go. It might seem odd coming from other languages.

Reply:
I'm not convinced that it's clearer. Main problem with this approach is that you have to take care not to add an extra separator at the end. Since we're using the comma as a separator to match with the rest of the formatting we'll end up with an extra comma then. 

Isn't `strings.Join(listOfStrings, ", ")` pretty clear? It does basically the same thing inside using a builder, and you know it'll handle placing the separator edge cases correctly.

Reply:
Sure. You can use the join. Performance isn’t critical here. It was just an example. Creating a slice and then joining will likely be the easiest.